scrapers/utils.py

```
@@ -134,3 +142,89 @@ def validate_and_get_download_link(base_url, download_link, check_content_pdf=Tr
             raise TypeError("File extension does not match content type.")
 
     return absolute_url
+
+
+def calculate_checksum(filepaths):
+    md5 = hashlib.md5()
+    for file in filepaths:
+        try:
+            md5.update(Path(file).read_bytes())
+        except IsADirectoryError:
+            pass
+    return md5.hexdigest()
+
+
+def zip_pdfs(zipped_directory, scraper_filepath, filepaths, link_id):
+    zip_name = get_standardized_filename(scraper_filepath, link_id, ".zip")
+    zip_path = get_file_path(zipped_directory, zip_name)
+
+    with zipfile.ZipFile(zip_path, "w") as zipf:
+        for file in filepaths:
+            filename = os.path.basename(file)
+            zipf.write(file, arcname=filename)
+
+    return zip_name, zip_path
+
+
+def connect_circulars_to_associated_files(file_names):
+    """
+    Accepts 'file_names' as an argument, which is a list of lists in which each sub-list consists of a file name at index 0 and a file extension at index 1.
+
+    This function identifies files that are part of the same circular and groups them together in a returned dictionary, in which each key is a circular name and its value is a list of all files belonging to that circular.
+
+    Example of valid input:     [ ["file1", "pdf"], ["file2", "pdf"], ["file3", "docx"] ]   OR    [ ("file1", "pdf") ]
+
+    Invalid input:      [ ["file1", 1] ]    OR    [ {"file1", "pdf"}, {"file2", "pdf"} ]    OR  { "file1": "pdf", "file2": "pdf" }
+    """
+    circular_to_links_dict = {}
+    try:
+        # sort file_names list by file name string length
+        file_names.sort(key=lambda x: len(x[0]))
+    except Exception as e:
+        raise TypeError(f"Invalid 'file_names' argument.")
```

I would prefer letting `sort` throw the actual errors here. This way we are not masking the actual errors, making debugging difficult